include/seqan3/alignment/matrix/alignment_trace_matrix.hpp

```diff
@@ -98,29 +99,108 @@ struct alignment_coordinate
+        if ((dir & L) == L)
+        {
+            col = std::max<signed_size_t>(col - 1, 0);
+            ++inserts;
+        }
+        else if ((dir & U) == U)
+        {
+            row = std::max<signed_size_t>(row - 1, 0);
+            ++deletes;
+        }
+        else if ((dir & D) == D)
+        {
+            row = std::max<signed_size_t>(row - 1, 0);
+            col = std::max<signed_size_t>(col - 1, 0);
+            ++substitutes;
+        }
+        else
+        {
+            if (row == 0 || col == 0)
+                break;
+            throw std::logic_error{"Trace not found"};
+        }
+    }
+    return {inserts, deletes, substitutes};
+}
+
+/*!\brief Compute the begin coordinate.
+ * \ingroup alignment_matrix
+ * \tparam    trace_matrix_t The type of the trace matrix.
+ * \param[in] end_coordinate Where the trace in the matrix ends.
+ * \param[in] statistics     Statistics returned by alignment_trace_statistics.
+ */
+inline alignment_coordinate
+alignment_begin_coordinate(alignment_coordinate const end_coordinate, std::tuple<size_t, size_t, size_t> const & statistics)
+{
+    using signed_size_t = std::make_signed_t<size_t>;
+    auto && [insertions, deletes, substitutes] = statistics;
+    return
+    {
+        static_cast<size_t>(std::max<signed_size_t>(end_coordinate.seq1_pos - insertions - substitutes, 0)),
+        static_cast<size_t>(std::max<signed_size_t>(end_coordinate.seq2_pos - deletes - substitutes, 0))
+    };
+}
+
+/*!\brief Compute the trace from a trace matrix.
+ * \ingroup alignment_matrix
+ * \tparam    database_t           The type of the database sequence.
+ * \tparam    query_t              The type of the query sequence.
+ * \tparam    trace_matrix_t       The type of the trace matrix.
+ * \param[in] database             The database sequence.
+ * \param[in] query                The query sequence.
+ * \param[in] matrix               The trace matrix.
+ * \param[in] end_coordinate       Where the trace in the matrix ends.
+ */
 template <
     typename database_t,
     typename query_t,
     typename trace_matrix_t,
-    typename gapped_database_t = gapped<std::decay_t<decltype(std::declval<database_t>()[0])>>,
-    typename gapped_query_t = gapped<std::decay_t<decltype(std::declval<query_t>()[0])>>>
+    typename gapped_database_alphabet_t = gapped<std::decay_t<decltype(std::declval<database_t>()[0])>>,
+    typename gapped_query_alphabet_t = gapped<std::decay_t<decltype(std::declval<query_t>()[0])>>>
 //!\cond
     requires matrix_concept<remove_cvref_t<trace_matrix_t>> &&
              std::Same<typename remove_cvref_t<trace_matrix_t>::entry_type, trace_directions>
 //!\endcond
-inline std::pair<std::list<gapped_database_t>, std::list<gapped_query_t>>
-alignment_trace(database_t && database, query_t && query, trace_matrix_t && matrix, alignment_coordinate const end_coordinate)
+inline std::pair<std::vector<gapped_database_alphabet_t>, std::vector<gapped_query_alphabet_t>>
+alignment_trace(
```

Comment:
please move the database parameter into this line and indent the rest accordingly